src/comps/qb.cc

```diff
@@ -0,0 +1,407 @@
+
+        // Updating B norm estimation
+        T norm_B_i = lange(Norm::Fro, block_sz, n, B_i_dat, block_sz);
+        norm_B = hypot(norm_B, norm_B_i);
+        // Updating approximation error
+        prev_err = approx_err;
+        approx_err = sqrt(abs(norm_A - norm_B) * (norm_A + norm_B)) / norm_A;
+
+        // Early termination - handling round-off error accumulation
+        if ((curr_sz > 0) && (approx_err > prev_err))
+        {
+            // Early termination - error growth
+            // Resizing the output arrays
+            RandLAPACK::comps::util::qb_resize( m, n, Q, B, k, curr_sz);
+            QB::termination = 2;
+            return;
+        } 
+
+        // Update the matrices Q and B
+        lacpy(MatrixType::General, m, block_sz, Q_i_dat, m, Q_dat + (m * curr_sz), m);	
+        lacpy(MatrixType::General, block_sz, n, B_i_dat, block_sz, B_dat + curr_sz, k);
+
+        //char name_3[] = "QB1  output Q";
+        //RandBLAS::util::print_colmaj(m, k, Q.data(), name_3);
+
+        //char name_4[] = "QB1  output B";
+        //RandBLAS::util::print_colmaj(k, n, B.data(), name_4);
+
+        if(QB::orth_check)
+        {
+            gemm(Layout::ColMajor, Op::Trans, Op::NoTrans,
+                k, k, m,
+                1.0, Q_dat, m, Q_dat, m,
+                0.0, Q_gram_dat, k
+            );
+            for (int oi = 0; oi < next_sz; ++oi) {
+                Q_gram_dat[oi * k + oi] -= 1.0;
+            }
+            T orth_err = lange(Norm::Fro, k, k, Q_gram_dat, k);
+    
+            if(QB::verbosity)
+            {
+               printf("Q ERROR:   %e\n\n", orth_err);
+            }
+
+            if (orth_err > 1.0e-10)
+            {
+                // Lost orthonormality of Q
+                // Cut off the last iteration
+                RandLAPACK::comps::util::qb_resize( m, n, Q, B, k, curr_sz);
+                QB::termination = 5;
+                return;
+            }
+        }
+
+        curr_sz += block_sz;
+        // Termination criteria
+        if (approx_err < tol)
+        {
+            // Expected ternimation - tolerance achieved
+            // Resizing the output arrays 
+            RandLAPACK::comps::util::qb_resize( m, n, Q, B, k, curr_sz);
+            QB::termination = 0;
+            return;
+        }
+        
+        // This step is only necessary for the next iteration
+        // A = A - Q_i * B_i
+        gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, n, block_sz, -1.0, Q_i_dat, m, B_i_dat, block_sz, 1.0, A_cpy_dat, m);
+
+        //char name_5[] = "A_updated";
+        //RandBLAS::util::print_colmaj(m, n, A_cpy_dat, name_5);
+    }
+    // Reached expected rank without achieving the tolerance
+    QB::termination = 3;
+}
+
+template <typename T>
+void QB<T>::QB2_test_mode(
+    int64_t m,
+    int64_t n,
+    std::vector<T>& A,
+    int64_t& k,
+    int64_t block_sz,
+    T tol,
+    std::vector<T>& Q,
+    std::vector<T>& B
+){
```

Comment:
My understanding is that this function is almost identical to ``QB2``. You should have comments that indicate where the differences are.